src/Common/src/System/Net/IPv6AddressHelper.Common.cs

```diff
@@ -7,5 +7,407 @@ namespace System
+            {
+                if (numbers[i] == 0)
+                {
+                    currentSequenceLength++;
+                    if (currentSequenceLength > longestSequenceLength)
+                    {
+                        longestSequenceLength = currentSequenceLength;
+                        longestSequenceStart = i - currentSequenceLength + 1;
+                    }
+                }
+                else
+                {
+                    currentSequenceLength = 0;
+                }
+            }
+
+            return longestSequenceLength > 1 ?
+                (longestSequenceStart, longestSequenceStart + longestSequenceLength) :
+                (-1, -1);
+        }
+
+        // Returns true if the IPv6 address should be formatted with an embedded IPv4 address:
+        // ::192.168.1.1
+        internal unsafe static bool ShouldHaveIpv4Embedded(ReadOnlySpan<ushort> numbers)
+        {
+            // 0:0 : 0:0 : x:x : x.x.x.x
+            if (numbers[0] == 0 && numbers[1] == 0 && numbers[2] == 0 && numbers[3] == 0 && numbers[6] != 0)
+            {
+                // RFC 5952 Section 5 - 0:0 : 0:0 : 0:[0 | FFFF] : x.x.x.x
+                if (numbers[4] == 0 && (numbers[5] == 0 || numbers[5] == 0xFFFF))
+                {
+                    return true;
+                }
+                // SIIT - 0:0 : 0:0 : FFFF:0 : x.x.x.x
+                else if (numbers[4] == 0xFFFF && numbers[5] == 0)
+                {
+                    return true;
+                }
+            }
+
+            // ISATAP
+            return numbers[4] == 0 && numbers[5] == 0x5EFE;
+        }
+
+        //
+        // IsValidStrict
+        //
+        //  Determine whether a name is a valid IPv6 address. Rules are:
+        //
+        //   *  8 groups of 16-bit hex numbers, separated by ':'
+        //   *  a *single* run of zeros can be compressed using the symbol '::'
+        //   *  an optional string of a ScopeID delimited by '%'
+        //   *  the last 32 bits in an address can be represented as an IPv4 address
+        //
+        //  Difference between IsValid() and IsValidStrict() is that IsValid() expects part of the string to 
+        //  be ipv6 address where as IsValidStrict() expects strict ipv6 address.
+        //
+        // Inputs:
+        //  <argument>  name
+        //      IPv6 address in string format
+        //
+        // Outputs:
+        //  Nothing
+        //
+        // Assumes:
+        //  the correct name is terminated by  ']' character
+        //
+        // Returns:
+        //  true if <name> is IPv6  address, else false
+        //
+        // Throws:
+        //  Nothing
+        //
+
+        //  Remarks: MUST NOT be used unless all input indexes are verified and trusted.
+        //           start must be next to '[' position, or error is reported
+        internal unsafe static bool IsValidStrict(char* name, int start, ref int end, bool allowAnyScope = false)
```

Comment:
We're not currently passing in true for allowAnyScope anywhere.  You referred to this change in your PR message.  Is it that you're passing true in from some Mono code?

You also mentioned a difference in what's supported between netfx and core.  Is there a test you could add that would highlight the difference?  It may just be a bug in core that we should fix. cc: @davidsh

Reply:
This is one of those Mono-specific hacks that I am going to remove as I just mentioned in my last comment.